server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageStats.java

```diff
@@ -87,7 +87,15 @@ private void checkByteSize(long bytes) {
 
     private PerFieldDiskUsage getOrAdd(String fieldName) {
         Objects.requireNonNull(fieldName, "fieldName must be non-null");
-        return fields.computeIfAbsent(fieldName, k -> new PerFieldDiskUsage());
+        try {
+            return fields.computeIfAbsent(fieldName, k -> new PerFieldDiskUsage());
+        } catch (UnsupportedOperationException e) {
+            /* This happens if this object had been serialized and deserialized while fields was empty, since StreamInput returns
+             * immutable maps for empty maps
+             */
+            this.fields = new HashMap<>();
```

Comment:
I think this is lossy, because if we were to ever change `StreamInput::readMap` to return `Collections.unmodifiableMap(<some-map-with-data>)` then it would be reset back to an empty one here and we would "lose" some of the data.

I think instead, because we know we might be changing it, we should do:

```java
this.fields = new HashMap<>(in.readMap(StreamInput::readString, PerFieldDiskUsage::new));
```

So that it is always available for changing. Then, in future work, we can look at making this class behave in a more immutable style.

Reply:
Hmm, good point. I'm assuming that this is called infrequently enough that the minor performance hit is not going to matter, right?